File: src/layer.c
#include "layer.h"

Layer create_layer(int num_neurons, int num_inputs_per_neuron)
{
    Layer layer = {
        .num_neurons = num_neurons,
        .neurons = malloc(num_neurons * sizeof(Neuron))};

    for (int i = 0; i < num_neurons; i++)
    {
        layer.neurons[i] = create_neuron(num_inputs_per_neuron);
    }

    return layer;
}

File: src/layer.rs
use crate::neuron::Neuron;

pub struct Layer {
    pub num_neurons: usize,
    pub neurons: Vec<Neuron>,
}

impl Layer {
    pub fn new(num_neurons: usize, num_inputs: usize) -> Self {
        let mut neurons = Vec::new();

        for _ in 0..num_neurons {
            neurons.push(Neuron::new(num_inputs));
        }

        Self {
            num_neurons,
            neurons,
        }
    }

    pub fn set_inputs(&mut self, inputs: &[f64]) {
        for neuron in &mut self.neurons {
            neuron.set_inputs(inputs);
        }
    }

    pub fn get_error(&mut self, neuron_index: usize) -> f64 {
        let mut error: f64 = 0.0;

        for neuron in self.neurons.iter() {
            error += neuron.weights[neuron_index] * neuron.error;
        }

        error
    }
}
